### PR TITLE
Support regex tags search for Elasticseach backend

### PR DESCRIPTION
**Pull request by annanay25**
Signed-off-by: Annanay <annanayagarwal@gmail.com>

## Which problem is this PR solving?
When using the ES backend, adds the ability to 
- use regex search for the values field. So a query like `foo=bar.*` should match all tags like `{foo:bar}, {foo:barbaz}, {foo:bar1}`

Related to #684 

## Short description of the changes
- Swapped the `NewMatchQuery` method in ES reader for `NewRegexpQuery`

/cc @pavolloffay

### CODE REVIEW

**Comment by annanay25** [comment]
I've added the "search for presence of a tag" functionality to the cassandra reader as well, but I can submit that in a different PR

**Comment by pavolloffay** [comment]
> search for presence of a tag

Any ideas how we could support this in UI? 

Could you please add integration tests for regex and also escaped version e.g. find exactly `hello*`.

**Comment by pavolloffay** [comment]
It would be better to split the PR for regex and missing tags to two separate PRs.

**Comment by annanay25** [comment]
@pavolloffay - Please review.

**Comment by yurishkuro** [comment]
What are the performance implications of always forcing a regex search?

**Comment by annanay25** [comment]
@yurishkuro From what I understand, Elasticsearch performance degrades with decrease in prefix length before the wildcard(operator) in the regex query. For ex: `http.url=*.example.com` will have worse performance as compared to `http.url=example.c*` 

I could not find performance benchmarks for a Match query vs Regexp query for when no operator (*,?,etc) is used.

**Comment by pavolloffay** [comment]
I am also interested in performance results ad tag queries are common. Could you please run some perf tests to measure the query time?

I have used this https://github.com/pavolloffay/jaeger-perf-tests before and also @jkandasa should have some perf framework, although I am not sure if that can be used without our infrastructure.

**Comment by pavolloffay** [comment]
If there is perf degradation maybe we could use regex type only if there is regex in the query.

**Comment by jkandasa** [comment]
@annanay25 we have a tool to measure query timings. https://github.com/jkandasa/jaegerperf

**Comment by annanay25** [comment]
@pavolloffay - Thanks. Is there a guide on which metrics need to be monitored after running the perf-tests?

@jkandasa - Could you please provide some more information on the tool? Does it perf-test the collector/query component, etc? The README doesn't say much.

**Comment by annanay25** [comment]
@pavolloffay - Also, in the integration test, the "Tag Name + Operation Name" query seems to match the `tags_regex_trace.json` as well, even though they search on a different `ServiceName` and tag value. 

I'm not sure I understand why that is happening, could you please take a look?

File: plugin/storage/integration/fixtures/queries.json [hunk 1]
```diff
@@ -376,5 +376,37 @@
...
+      "StartTimeMin": "2017-01-26T15:46:31.639875Z",
+      "StartTimeMax": "2017-01-26T17:46:31.639875Z",
+      "DurationMin": 0,
+      "DurationMax": 1000,
+      "NumTraces": 1000
+    },
+    "ExpectedFixtures": ["tags_regex_trace"]
+  },
+  {
+    "Caption": "Tag regex + Operation name + max Duration - Multiple traces",
```
Comment (pavolloffay) [inline plugin/storage/integration/fixtures/queries.json]: > @pavolloffay - Also, in the integration test, the "Tag Name + Operation Name" query seems to match the tags_regex_trace.json as well, even though they search on a different ServiceName and tag value.

@annanay25 do you mean this one?

Comment (annanay25) [reply plugin/storage/integration/fixtures/queries.json]: No, the one above, "Tag regex + Operation name + max Duration"

It is supposed to match with just the `tags_regex_trace` but matches with both `tags_regex_trace` and `tags_opname_trace `.

Comment (pavolloffay) [reply plugin/storage/integration/fixtures/queries.json]: This one also looks wrong as it uses different service name https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/integration/fixtures/traces/tags_opname_trace.json

Could you please debug it and or open an issue?

**Comment by jkandasa** [comment]
> @jkandasa - Could you please provide some more information on the tool? Does it perf-test the collector/query component, etc? The README doesn't say much.

[This tool ](https://github.com/jkandasa/jaegerperf)can send spans to the collector component. and can be used to measure jaeger-query API execution time. Both actions are independent.
I have updated the README file. Let me know still it is not clear.

File: plugin/storage/integration/fixtures/queries.json (hunk 1 above)
Comment (pavolloffay) [reply plugin/storage/integration/fixtures/queries.json]: How did you fix the issue?

Comment (annanay25) [reply plugin/storage/integration/fixtures/queries.json]: I had to mangle the traceID further - 
https://github.com/jaegertracing/jaeger/pull/2049/commits/0955efa38d658f508c64c41f71c5eafdbdf0d8ac#diff-1824d3ce6ce534210f34269d2370e966

**Comment by annanay25** [comment]
@jkandasa - I was able to setup your tool to do the performance tests. I loved the simplicity, and the concept of creating jobs. I have a few suggestions from my experience of using it - 

- It might be helpful to mention somewhere in the `README` that the QueryRunner will generate a JSON with results once it completes. It was not obvious to me, at first
- It would also be useful to have some guidelines on generated-jsons can be compared to get an idea of performance
- In the Jobs UI page, clicking on the "delete job" icon shows "Status: Completed" even for a running job. 
- I'm also attaching the [kubernetes yaml](https://gist.github.com/annanay25/454744eaea91fbde1d7b9ff5c765e0f3) I wrote to set this up - please feel free to include it in the repo if you wish as it might be useful to others.

**Comment by annanay25** [comment]
Perf results:

From master build -
```
{
    "summary": [
     {
      "contentLength": 242,
      "elapsed": 4388,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "3.last 2 days limit 1000",
      "samples": 10
     },
     {
      "contentLength": 304,
      "elapsed": 5092,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "4.last 7 days limit 2000",
      "samples": 10
     },
     {
      "contentLength": 282153,
      "elapsed": 203237,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "1.last 12 hours limit 100",
      "samples": 10
     },
     {
      "contentLength": 2923946,
      "elapsed": 2138975,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "1.last 12 hours limit 1000",
      "samples": 10
     },
     {
      "contentLength": 2970425,
      "elapsed": 21746050,
      "errorPercentage": 50,
      "errorsCount": 5,
      "name": "1.last 12 hours limit 2000",
      "samples": 10
     },
     {
      "contentLength": 229,
      "elapsed": 93426,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "2.last 24 hours limit 100",
      "samples": 10
     },
     {
      "contentLength": 229,
      "elapsed": 81108,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "2.last 24 hours limit 1000",
      "samples": 10
     },
     {
      "contentLength": 229,
      "elapsed": 7399,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "2.last 24 hours limit 2000",
      "samples": 10
     },
     {
      "contentLength": 242,
      "elapsed": 4355,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "3.last 2 days limit 100",
      "samples": 10
     },
     {
      "contentLength": 242,
      "elapsed": 4051,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "3.last 2 days limit 2000",
      "samples": 10
     },
     {
      "contentLength": 304,
      "elapsed": 4632,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "4.last 7 days limit 100",
      "samples": 10
     },
     {
      "contentLength": 304,
      "elapsed": 4352,
      "errorPercentage": 100,
      "errorsCount": 10,
      "name": "4.last 7 days limit 1000",
      "samples": 10
     }
    ]
   }

```

With this PR - 
```
{
    "summary": [
     {
      "contentLength": 3546187,
      "elapsed": 2750042,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "3.last 2 days limit 2000",
      "samples": 10
     },
     {
      "contentLength": 232174,
      "elapsed": 997305,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "1.last 12 hours limit 100",
      "samples": 10
     },
     {
      "contentLength": 2609034,
      "elapsed": 10571863,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "1.last 12 hours limit 1000",
      "samples": 10
     },
     {
      "contentLength": 3546187,
      "elapsed": 12699653,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "1.last 12 hours limit 2000",
      "samples": 10
     },
     {
      "contentLength": 2609034,
      "elapsed": 2888810,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "2.last 24 hours limit 1000",
      "samples": 10
     },
     {
      "contentLength": 3546187,
      "elapsed": 4955535,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "2.last 24 hours limit 2000",
      "samples": 10
     },
     {
      "contentLength": 230222,
      "elapsed": 571004,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "3.last 2 days limit 100",
      "samples": 10
     },
     {
      "contentLength": 2609034,
      "elapsed": 1898116,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "3.last 2 days limit 1000",
      "samples": 10
     },
     {
      "contentLength": 230222,
      "elapsed": 379544,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "4.last 7 days limit 100",
      "samples": 10
     },
     {
      "contentLength": 2609034,
      "elapsed": 1753636,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "4.last 7 days limit 1000",
      "samples": 10
     },
     {
      "contentLength": 3546187,
      "elapsed": 2979869,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "4.last 7 days limit 2000",
      "samples": 10
     },
     {
      "contentLength": 230222,
      "elapsed": 1225345,
      "errorPercentage": 0,
      "errorsCount": 0,
      "name": "2.last 24 hours limit 100",
      "samples": 10
     }
    ]
   }
```


Bunch of the test results from the master build give an `errorPercentage: 100`

**Comment by jkandasa** [comment]
@annanay25 Thank you for the comments and feedback on the tool. I have added a metrics page to compare the results
.
> * It might be helpful to mention somewhere in the `README` that the QueryRunner will generate a JSON with results once it completes. It was not obvious to me, at first

Updated.

> * It would also be useful to have some guidelines on how two generated-jsons can be compared to get an idea of performance

Metrics page includeded

> * In my latest run I get a constant contentLength of 56 in all summary results, seems fishy, is there an explanation?

In this case, you may see errors or no data for the specified timerange. Please refer JSON result. ` metrics >> raw`
Example:
```
{
       "contentLength": 194,
       "elapsed": 15166842,
       "errorMessage": "",
       "others": {
        "errors": [
         {
          "code": 500,
          "msg": "Get https://elasticsearch:9200/_msearch: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"
}
```

> * In the Jobs UI page, clicking on the "delete job" icon shows "Status: Completed" even for a running job.

Fixed

> * I'm also attaching the [kubernetes yaml](https://gist.github.com/annanay25/454744eaea91fbde1d7b9ff5c765e0f3) I wrote to set this up - please feel free to include it in the repo if you wish as it might be useful to others.

Thank you. I have modified and added in the source code.

**Comment by annanay25** [comment]
Seems like there is some performance degradation, how do you suggest this can be improved @pavolloffay?

**Comment by pavolloffay** [comment]
@annanay25 could you please create a summary of the test results? For instance what types of queries take how much time in % against master/head https://github.com/jaegertracing/jaeger/pull/1969#issuecomment-579695719

**Comment by annanay25** [comment]
@pavolloffay - https://docs.google.com/spreadsheets/d/1mzJNQYv1xF0fiMAcpXk5TlNVE3GIqZOlZv2iDZqkraI/edit?usp=sharing

On my system, after one run the perf tests start giving a 100% error for most queries in the master build (1.16) - can someone else test this?

**Comment by pavolloffay** [comment]
@annanay25 the results do not look good, the performance degradation is quite significant.

I would suggest detecting the regex and changing the query type on the fly only when it is necessary.

**Comment by annanay25** [comment]
@pavolloffay - Tests on my machine are quite inconsistent (PTAL at round 2, numbers are in complete contrast with round 1)

Can someone else run a benchmark on their machines? I can publish my image on dockerhub for this

**Comment by pavolloffay** [comment]
@annanay25 could you please build the query and collector image and publish it? @jkandasa would you be able to run the same tests as in #1969 ?

**Comment by jkandasa** [comment]
@pavolloffay ok I will do.
@annanay25 can you publish the collector and query image?

**Comment by annanay25** [comment]
Thanks @jkandasa. 
For collector the `jaegertracing/jaeger-collector` image can be used, as there are no changes for the collector in this PR. 
For query, I have pushed the image at https://hub.docker.com/r/annanay25/jaeger-query

**Comment by smpavlenko** [comment]
Hello,
is there any ETA when it will be released?

**Comment by annanay25** [comment]
@jkandasa - Did you get a chance to take a look at this?

**Comment by jkandasa** [comment]
@annanay25 here is the result:

image

Detailed file: https://docs.google.com/spreadsheets/d/1fIOYd4LJHHcmOIlEYuR9TQlAv4SXWdddy22zPmymvvw/edit?usp=sharing

##### Image Details:
* 1.17.0
```
Collecotor: jaegertracing/jaeger-collector:1.17.0
Query: jaegertracing/jaeger-query:1.17.0
```
* pr #2049 
```
Collector: jaegertracing/jaeger-collector:1.17.0
Query: annanay25/jaeger-query:latest (docker.io/annanay25/jaeger-query@sha256:3d700db8fb31b43cced6ecbfb7c17452165fa3def0525b9e7b366b90de9f1f6d)
```
* OCP version
```
Server Version: 4.3.1
Kubernetes Version: v1.16.2
```

Test data:
```
target: "collector"
endpoint: http://jaegerqe-collector:14268/api/traces
serviceName: jaegerperf_generator
mode: history
numberOfDays: 10
spansPerDay: 60000
spansPerSecond: 300 # maximum spans limit/sec
childDepth: 2
tags: 
  spans_generator: "jaegerperf"
  days: 10
startTime: 
```

##### Query test raw data:
* [query_test_raw_data.zip](https://github.com/jaegertracing/jaeger/files/4275133/query_test_raw_data.zip)

**Comment by annanay25** [comment]
Thank you for the results @jkandasa.

@pavolloffay - Could you please review these?

File: plugin/storage/integration/elasticsearch_test.go [hunk 2]
```diff
@@ -96,6 +96,8 @@ func (s *ESStorageIntegration) initializeES(allTagsAsFields, archive bool) error
 	s.esCleanUp(allTagsAsFields, archive)
 	// TODO: remove this flag after ES support returning spanKind when get operations
 	s.notSupportSpanKindWithOperation = true
+
+	s.Fixtures = loadAndParseQueryTestCases(t, "fixtures/queries_es.json")
```
Comment (pavolloffay) [inline plugin/storage/integration/elasticsearch_test.go]: It's better to load them in `testElasticsearchStorage`. This func initializes the storage.

File: plugin/storage/integration/fixtures/queries_es.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+[
+    {
+        "Caption": "Tag excaped operator + Operation name + max Duration",
```
Comment (pavolloffay) [inline plugin/storage/integration/fixtures/queries_es.json]: s/excaped/escaped/

File: plugin/storage/integration/fixtures/queries_es.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        "Query": {
+        "ServiceName": "query23-service",
+        "OperationName": "query23-operation",
+        "Tags": {
+            "sameplacetag1":"random\\*"
```
Comment (pavolloffay) [inline plugin/storage/integration/fixtures/queries_es.json]: Can we use `same\\*`. 

The other fixtures contain `sameplacetag1: sameplacevalue` so we should see it does not bring other results.

File: plugin/storage/integration/fixtures/queries_es.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        "StartTimeMin": "2017-01-26T15:46:31.639875Z",
+        "StartTimeMax": "2017-01-26T17:46:31.639875Z",
+        "DurationMin": 0,
+        "DurationMax": 1000,
+        "NumTraces": 1000
+        },
+        "ExpectedFixtures": ["tags_escaped_operator_trace"]
+    },
+    {
+        "Caption": "Tag wildcard regex + Operation name + max Duration",
```
Comment (pavolloffay) [inline plugin/storage/integration/fixtures/queries_es.json]: Operation name is empty

File: plugin/storage/integration/fixtures/queries_es.json [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        "Tags": {
+            "sameplacetag1":"same.*"
+        },
+        "StartTimeMin": "2017-01-26T15:46:31.639875Z",
+        "StartTimeMax": "2017-01-26T17:46:31.639875Z",
+        "DurationMin": 0,
+        "DurationMax": 0,
+        "NumTraces": 1000
+        },
+        "ExpectedFixtures": ["tags_opname_trace"]
```
Comment (pavolloffay) [inline plugin/storage/integration/fixtures/queries_es.json]: Could we have multiple traces that match this query?

**Comment by pavolloffay** [comment]
@annanay25 could you please provide more description in the first comment? E.g. what fields could be searched with regex. Is it only values or the keys? 

Does it work with the our UI?

**Comment by annanay25** [comment]
Thanks for the review, @pavolloffay. I've addressed all comments. PTAL.

**Comment by pavolloffay** [comment]
> should match all tags like

what about logs?

File: plugin/storage/integration/elasticsearch_test.go [hunk 7]
```diff
@@ -68,7 +68,7 @@ func (s *ESStorageIntegration) getVersion() (uint, error) {
 	return uint(esVersion), nil
 }
 
-func (s *ESStorageIntegration) initializeES(allTagsAsFields, archive bool) error {
+func (s *ESStorageIntegration) initializeES(t *testing.T, allTagsAsFields, archive bool) error {
```
Comment (pavolloffay) [inline plugin/storage/integration/elasticsearch_test.go]: `t` can be removed

File: plugin/storage/integration/elasticsearch_test.go [hunk 8]
```diff
@@ -96,6 +96,7 @@ func (s *ESStorageIntegration) initializeES(allTagsAsFields, archive bool) error
 	s.esCleanUp(allTagsAsFields, archive)
 	// TODO: remove this flag after ES support returning spanKind when get operations
 	s.notSupportSpanKindWithOperation = true
+
```
Comment (pavolloffay) [inline plugin/storage/integration/elasticsearch_test.go]: nit: revert

**Comment by namaiiee2222** [comment]
Is there any update on having regex in tag search?

**Pull request merged**